src/base/http/basicrequesthandler.cpp

```diff
@@ -0,0 +1,56 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2021  Prince Gupta <guptaprince8832@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "basicrequesthandler.h"
+
+#include <QDateTime>
+#include <QObject>
+
+#include "base/utils/gzip.h"
+#include "connection.h"
+#include "httputils.h"
+
+void Http::BasicRequestHandler::handleRequest(const Request &request, Connection *connection)
+{
+    Response response {processRequest(request, connection->enviroment())};
+
+    if (acceptsGzipEncoding(request.headers[QLatin1String {"accept-encoding"}]))
+        response.headers[HEADER_CONTENT_ENCODING] = QLatin1String {"gzip"};
+
+    response.headers[HEADER_CONNECTION] = QLatin1String {"keep-alive"};
+    tryCompressContent(response);
+    if (!response.content.isEmpty())
+        response.headers[HEADER_CONTENT_LENGTH] = QString::number(response.content.size());
+    if (!response.headers.contains(HEADER_DATE))
+        response.headers.insert(HEADER_DATE, httpDate());
+
+    connection->sendStatus(response.status);
+    connection->sendHeaders(response.headers);
+    if (!response.content.isEmpty())
+        connection->sendContent(response.content);
```

Comment:
So why not serialize/write response at once (as it was before)?

Reply:
in https://github.com/qbittorrent/qBittorrent/pull/12542 I send content in chunks, so to support that I've added `sendHeaders` and `sendContent` functions in Connection class. I can add sendResponse function in the Connection class but let's keep `Response` handling in this class.

Reply:
>let's keep Response handling in this class.

What's the reason? Connection acts on HTTP layer. It parses received HTTP requests into Request object so why it shouldn't make an opposite action, i.e. serialize Response object? In addition, it will give more control over its integrity and consistency.
>in #12542 I send content in chunks

Is it really useful? IIRC, your tests show it is not so optimal.

Reply:
> > let's keep Response handling in this class.
> 
> What's the reason? Connection acts on HTTP layer. It parses received HTTP requests into Request object so why it shouldn't make an opposite action, i.e. serialize Response object? In addition, it will give more control over its integrity and consistency.
> 

connection class already provide an interface for sending data, why pollute it by providing other ways to do the same. As far as the integrity of responses is concerned. I don't want to add such logic in Connection class. I would like to keep it barebones and have the request handler handle everything.

> > in #12542 I send content in chunks
> 
> Is it really useful? IIRC, your tests show it is not so optimal.

in #12542 client can ask for 1 gig of file. would you like to send all that at once? with the latest revision, this PR uses a cache for sending headers and status this makes it on par with the current master in terms of performance.

Reply:
>in #12542 client can ask for 1 gig of file.

Are you going to accept such request?

>would you like to send all that at once? 

I assume you don't, but then how do you come back later after sending partial of it?

Reply:
> Are you going to accept such request?

yep, it's a valid request

>  I assume you don't, but then how do you come back later after sending partial of it?

this functionality is basically to allow https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests on downloading torrent files, I use libtorrent::read_piece which is asynchronous so there is a natural callback there. So we basically fulfill HTTP range requests in size of torrent piece length using libtorrent::read_piece

Reply:
>I use libtorrent::read_piece which is asynchronous so there is a natural callback there. 

Excuse me I haven't got time to check your code, so how do you handle `libtorrent::read_piece()` alert that will return in unspecified/random order? Do you issue only one read_piece() at a time?

>this functionality is basically to allow https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests 

Some more questions, which of the following do you intend to support?
* `Accept-Ranges: bytes` (this is the minimum requirement right?)
* `Accept-Ranges: none`
* Single part ranges
* Multipart ranges
* Conditional range requests

I hope we can aim to support the smallest set that will satisfy your need.

Reply:
> how do you handle libtorrent::read_piece() alert that will return in unspecified/random order? Do you issue only one read_piece() at a time?

yes, read_piece() is issued one at a time.

> which of the following do you intend to support?

only Single Part ranges.